src/audio_core/voice_context.cpp

```diff
@@ -0,0 +1,531 @@
+// Copyright 2020 yuzu Emulator Project
+// Licensed under GPLv2 or any later version
+// Refer to the license.txt file included.
+
+#include "audio_core/behavior_info.h"
+#include "audio_core/voice_context.h"
+#include "core/memory.h"
+
+namespace AudioCore {
+
+ServerVoiceChannelResource::ServerVoiceChannelResource(s32 id) : id(id) {}
+ServerVoiceChannelResource::~ServerVoiceChannelResource() = default;
+
+bool ServerVoiceChannelResource::InUse() const {
+    return in_use;
+}
+
+float ServerVoiceChannelResource::GetCurrentMixVolumeAt(std::size_t i) const {
+    ASSERT(i < AudioCommon::MAX_MIX_BUFFERS);
+    return mix_volume.at(i);
+}
+
+float ServerVoiceChannelResource::GetLastMixVolumeAt(std::size_t i) const {
+    ASSERT(i < AudioCommon::MAX_MIX_BUFFERS);
+    return last_mix_volume.at(i);
+}
+
+void ServerVoiceChannelResource::Update(VoiceChannelResource::InParams& in_params) {
+    in_use = in_params.in_use;
+    // Update our mix volumes only if it's in use
+    if (in_params.in_use) {
+        std::copy(in_params.mix_volume.begin(), in_params.mix_volume.end(), mix_volume.begin());
+    }
+}
+
+void ServerVoiceChannelResource::UpdateLastMixVolumes() {
+    std::copy(mix_volume.begin(), mix_volume.end(), last_mix_volume.begin());
```

Comment:
```suggestion
    last_mix_volume = mix_volume;
```

I'd suggest evaluating other instances of std::copy if they can be simplified as well.

There's no need for std::copy between arrays of the same size and contained type. `operator=` can handle this just fine.

Similarly for memset, which in most cases I've seen so far, can either be replaced with std::array's or `<algorithm>`'s fill.